### PR TITLE
Partition patient interceptor

### PR DESCRIPTION
**Pull request by jamesagnew**

### CODE REVIEW

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+		createPatientA();
+
+		Observation obs = new Observation();
+		obs.getSubject().setReference("Patient/A");
+		Long id = myObservationDao.create(obs).getId().getIdPartAsLong();
+
+		runInTransaction(() -> {
+			ResourceTable observation = myResourceTableDao.findById(id).orElseThrow(() -> new IllegalArgumentException());
+			assertEquals("Observation", observation.getResourceType());
+			assertEquals("A".hashCode(), observation.getPartitionId().getPartitionId());
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:91]: I know it works for this case, but should we be doing `"A".hashCode() % 15000` like in the actual interceptor? Subsequent test-writers may not pick as convenient patient IDs.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+	@Test
+	public void testCreateOrganization_ValidMembershipInCompartment() {
+		Organization org = new Organization();
+		org.setName("Foo");
+		Long id = myOrganizationDao.create(org).getId().getIdPartAsLong();
+
+		runInTransaction(() -> {
+			ResourceTable observation = myResourceTableDao.findById(id).orElseThrow(() -> new IllegalArgumentException());
+			assertEquals("Organization", observation.getResourceType());
+			assertEquals(-1, observation.getPartitionId().getPartitionId().intValue());
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:107]: can we constant out `-1` into a class variable for this test? Maybe `DEFAULT_PARTITION_ID`

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java [hunk 3]
```diff
@@ -0,0 +1,257 @@
...
+	}
+
+	private void createPatientA() {
+		Patient patient = new Patient();
+		patient.setId("Patient/A");
+		patient.setActive(true);
+		myPatientDao.update(patient);
+	}
+
+}
```
Comment (tadgh) [inline hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:257]: I have pushed two additional test cases to this branch, to demonstrate things that currently fail, but _shouldnt_, and things that don't fail, but _should_

**Review by tadgh — APPROVED** [review]
Approved, but with reading how threaded through all the code the idea of the different partitioning strategies is, it is probably worthwhile to break a bunch of the functionality into a partitioning strategy interface. 

Couple notes: 
1. I have pushed some code to this branch with failing tests to illustrate the minor concerns i have. 
2. Nice!

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java [hunk 4]
```diff
@@ -1211,26 +1209,39 @@ public String getCurrentVersionId(IIdType theReferenceElement) {
...
+		Set<Integer> readPartitions = null;
+		if (requestPartitionId.isAllPartitions()) {
+			entity = myEntityManager.find(ResourceTable.class, pid.getIdAsLong());
+		} else {
+			readPartitions = myRequestPartitionHelperService.toReadPartitions(requestPartitionId);
+			if (readPartitions.size() == 1) {
+				if (readPartitions.contains(null)) {
+					entity = myResourceTableDao.readByPartitionIdNull(pid.getIdAsLong()).orElse(null);
+				} else {
+					entity = myResourceTableDao.readByPartitionId(readPartitions.iterator().next(), pid.getIdAsLong()).orElse(null);
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/dao/BaseHapiFhirResourceDao.java:1225]: This is only hit if the `size() == 1`  (see line 1221) so the `iterator.next()` is just an easy way to get the only element out of a set.

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java [hunk 5]
```diff
@@ -260,6 +260,26 @@ public void testBundleComposition() {
 	}
 
 
+	@Test
+	public void testCreateInvalidUnquotedExtensionUrl() {
+		SearchParameter fooSp = new SearchParameter();
+		fooSp.setCode("foo");
+		fooSp.setType(Enumerations.SearchParamType.STRING);
+		fooSp.setTitle("FOO SP");
+		fooSp.setExpression("Patient.extension.where(url=http://foo).value");
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/dao/r4/FhirResourceDaoR4SearchCustomSearchParamTest.java:269]: done

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java (hunk 1 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:93]: oh, ya good call. done

File: hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptor.java [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+			}
+		} else {
+			IFhirPath fhirPath = myFhirContext.newFhirPath();
+			compartmentIdentity = compartmentSps
+				.stream()
+				.flatMap(param -> Arrays.stream(BaseSearchParamExtractor.splitPathsR4(param.getPath())))
+				.filter(path -> isNotBlank(path))
+				.map(path -> fhirPath.evaluateFirst(theResource, path, IBaseReference.class))
+				.filter(refOpt -> refOpt.isPresent())
+				.map(refOpt -> refOpt.get())
```
[the remark this answers is not in the document]
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/main/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptor.java:103]: done

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java (hunk 3 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:287]: good call on them, i've made them pass

File: hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java (hunk 2 above)
Comment (jamesagnew) [reply hapi-fhir-jpaserver-base/src/test/java/ca/uhn/fhir/jpa/interceptor/PatientIdPartitionInterceptorTest.java:109]: done

**Pull request merged**